### DISCUSSION TITLE
morphTo relationship issues when rehydrating model

### DISCUSSION THREAD

**Discussion by DePalmo** [livewire/livewire#4947] [7. Bugs (archived)] [answered]
Background
I have an app that uses `morphTo` relationship between three models. Main model is `Project` and has `morphTo` to two other models `Group` and `User`. In practice this means that a single project can be assigned either to `Group` or `User`. Both of these two models have their own, different relationships. For example, `User` can belong to one `Group`, while `Group` can have many `User`.

Bug
I have set a form where user can dynamically assign the `Project` either to `Group` or `User`. When I assign a `Project` to `Group` it all works well. When I assign a `Project` to `User` it still works well. But when I then assign the same `Project` back to `Group`, it fails with error `Call to undefined relationship [groups] on model [App\Modules\User\Entities\Group]`. Naturally, model `Group` does not have relationship of `groups`, but `User` does.
When I examined the request payload, I noticed that, for unknown reason, it submits the relations too, like so:
```
{
    "serverMemo":{
        "dataMeta":{
            "models":{
                "Project":{
                    "class":"App\\Entities\\Base\\Project\\Project",
                    "id":53,
                    "relations":[
                        "number",
                        "assigned",
                        "assigned.groups",
                        "assigned.language",
                        "assigned.roles"
                    ],
                    "connection":"mysql"
                }
            }
        },
    }
}
```
And got wondering, why? So this is my first question. Why submit these relationships back to server and what's wrong if they are not there?

Because when I was trying to debug this, I managed to fix the issue by skipping the submitted relationships by changing how the model is hydrated here:
- `Livewire\HydrationMiddleware\HydratePublicProperties::hydrateModel()`
- L144: `new ModelIdentifier($serialized['class'], $serialized['id'], [], $serialized['connection'])`

Notice that there's missing the `$serialized['relations']`, which was replaced by `[]`. I haven't done any extensive tests yet, so would appreciate a comment from people who have more experience with Livewire why are these relationships required to be submitted from front-end back to server.
And also if my solution is viable. Yes I know, not ideal, but I have tried `->refresh()` but it's not working since it receives the non-existing relationship from frontend and tries to load it.

**Comment by joshhanley** [comment, marked as answer]
@DePalmo yeah Livewire has to keep track of, and reload relations to ensure the properties are rehydrated to what they were in the previous request. If Livewire doesn't, that could potentially cause performance issues for people expecting the relations to already be loaded.

Livewire does make use of Laravel's serialiser under the hood, and I know it can have issues at times with polymorphic relations.

You could unset all the relations yourself in a dehydrate hook, and then reload them manually if you need them on rehydration.

To unset on dehydration of your project property you'd do something like this:
```php
public function dehydrateProject()
{
    $this->project->unsetRelations();
}
```

Hope this helps!

**Reply by DePalmo** [reply]
Sorry for my late response, it does help and it solved my problem for now. Will keep in mind the performance issue if I will ever notice it in the future.
